group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -0,0 +1,803 @@
+                throw new KafkaException("append failed.");
+            }
+        }
+    }
+
+    /**
+     * A simple Coordinator implementation that stores the records into a set.
+     */
+    private static class MockCoordinator implements Coordinator<String> {
+        private final TimelineHashSet<String> records;
+
+        MockCoordinator(
+            SnapshotRegistry snapshotRegistry
+        ) {
+            records = new TimelineHashSet<>(snapshotRegistry, 0);
+        }
+
+        @Override
+        public void replay(String record) throws RuntimeException {
+            records.add(record);
+        }
+
+        Set<String> records() {
+            return Collections.unmodifiableSet(new HashSet<>(records));
+        }
+    }
+
+    /**
+     * A CoordinatorBuilder that creates a MockCoordinator.
+     */
+    private static class MockCoordinatorBuilder implements CoordinatorBuilder<MockCoordinator, String> {
+        private SnapshotRegistry snapshotRegistry;
+
+        @Override
+        public CoordinatorBuilder<MockCoordinator, String> withSnapshotRegistry(
+            SnapshotRegistry snapshotRegistry
+        ) {
+            this.snapshotRegistry = snapshotRegistry;
+            return this;
+        }
+
+        @Override
+        public MockCoordinator build() {
+            return new MockCoordinator(Objects.requireNonNull(this.snapshotRegistry));
+        }
+    }
+
+    /**
+     * A CoordinatorBuilderSupplier that returns a MockCoordinatorBuilder.
+     */
+    private static class MockCoordinatorBuilderSupplier implements CoordinatorBuilderSupplier<MockCoordinator, String> {
+        @Override
+        public CoordinatorBuilder<MockCoordinator, String> get() {
+            return new MockCoordinatorBuilder();
+        }
+    }
+
+    @Test
+    public void testScheduleLoading() {
+        MockCoordinatorLoader loader = mock(MockCoordinatorLoader.class);
+        MockPartitionWriter writer = mock(MockPartitionWriter.class);
+        MockCoordinatorBuilderSupplier supplier = mock(MockCoordinatorBuilderSupplier.class);
+        MockCoordinatorBuilder builder = mock(MockCoordinatorBuilder.class);
+        MockCoordinator coordinator = mock(MockCoordinator.class);
+
+        CoordinatorRuntime<MockCoordinator, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinator, String>()
+                .withLoader(loader)
+                .withEventProcessor(new MockEventProcessor())
+                .withPartitionWriter(writer)
+                .withCoordinatorBuilderSupplier(supplier)
+                .build();
+
+        when(builder.withSnapshotRegistry(any())).thenReturn(builder);
+        when(builder.build()).thenReturn(coordinator);
+        when(supplier.get()).thenReturn(builder);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        when(loader.load(TP, coordinator)).thenReturn(future);
+
+        // Getting the coordinator context fails because the coordinator
```

Comment:
does this fail not because we aren't loaded, but because we haven't even started loading? On 201, it says we are still loading, but were able to get the context.

Reply:
In this case, it fails because the context does not even exist.

Reply:
Got it -- the comment is a little unclear, since we can get the context while it is still not fully loaded, but the issue is that we didn't start loading (which starts with creating the context). 

Maybe we could say "Getting the coordinator context fails because the coordinator hasn't started loading and hasn't crated the context"? That might be a bit wordy, but something like that.

Reply:
Updated the comment.